formula-android/src/main/java/com/instacart/formula/android/internal/AndroidUpdateScheduler.kt

```diff
@@ -0,0 +1,70 @@
+package com.instacart.formula.android.internal
+
+import java.util.concurrent.atomic.AtomicBoolean
+import java.util.concurrent.atomic.AtomicReference
+
+/**
+ * Handles state update scheduling to the main thread. If update arrives on a background thread,
+ * it will added it the main thread queue. It will throw away a pending update if a new update
+ * arrives.
+ */
+class AndroidUpdateScheduler<Value : Any>(
+    private val update: (Value) -> Unit,
+) {
+    /**
+     * If not null, that means that we have an update pending.
+     */
+    private val pendingValue = AtomicReference<Value>()
+
+    /**
+     * Defines if an update is currently scheduled.
+     */
+    private val updateScheduled = AtomicBoolean(false)
+
+    /**
+     * To avoid re-entry, we track if [updateRunnable] is currently handling an update.
+     */
+    private var isUpdating = false
+
+    private val updateRunnable = object : Runnable {
+        override fun run() {
+            updateScheduled.set(false)
+
+            var localPending = pendingValue.getAndSet(null)
+            while (localPending != null) {
+                // Handle the update
+                isUpdating = true
+                update(localPending)
+                isUpdating = false
+
+                // Check if another update arrived while we were processing.
+                localPending = pendingValue.getAndSet(null)
+
+                if (localPending != null) {
+                    // We will take over processing, so let's clear the message
+                    Utils.mainThreadHandler.removeCallbacks(this)
```

Comment:
I think we also want to set updateScheduled to false here as well? If emitUpdate was called during the above update, the `pendingValue` would be set and updateScheduled would have been set to true.

Reply:
Good catch, moved it after `localPending != null)`